src/lib/statistics/chunk_statistics/counting_quotient_filter.hpp

```diff
@@ -0,0 +1,58 @@
+#pragma once
+
+#include <vector>
+#include <string>
+
+#include "cqf2.hpp"
+#include "cqf4.hpp"
+#include "cqf8.hpp"
+#include "cqf16.hpp"
+#include "cqf32.hpp"
+#include "types.hpp"
+
+#include "storage/base_segment.hpp"
+#include "abstract_filter.hpp"
+
+
+
+
+namespace opossum {
+
+/* Counting Quotient Filters allow you to keep track of which values are present in a segment and how often. Filters
+work approximately. If a membership query yields a positive result, the value is probably present but there is
+a chance of a false positive. If the query delivers a negative result, the item is guaranteed to not be contained.
+In the same way, items can be over counted but not under counted.
+CQF can be configured with quotient size, which determines the number of slots, and the remainder size, which
+corresponds to the slot size. At this time, the remainder size must be 2, 4, 8, 16 or 32. */
+
+template <typename ElementType>
+class CountingQuotientFilter : AbstractFilter {
+ public:
+  CountingQuotientFilter(uint8_t quotient_size, uint8_t remainder_size);
+  virtual ~CountingQuotientFilter();
+  CountingQuotientFilter(const CountingQuotientFilter& filter) = delete;
+  CountingQuotientFilter(CountingQuotientFilter&& filter) = default;
+  CountingQuotientFilter& operator=(const CountingQuotientFilter& filter) = delete;
+  CountingQuotientFilter& operator=(CountingQuotientFilter&& filter) = default;
+
+  void insert(ElementType value, uint64_t count);
+  void insert(ElementType value);
+  void populate(const std::shared_ptr<const BaseSegment>& segment);
+  uint64_t count(ElementType value) const;
+  uint64_t count_all_type(AllTypeVariant value) const;
+  uint64_t memory_consumption() const;
+  double load_factor() const;
+  bool is_full() const;
+
+  bool can_prune(const AllTypeVariant& value, const PredicateCondition predicate_type) const override;

```

Comment:
Same as the discussion in #1118. I believe that "dumb" data structures should not know about high-level database concepts, such as pruning. This might just be a naming discussion, but the only thing that the filter should tell is "Can I guarantee that this value is (not) contained in the segment?". Translating this to a pruning decision should happen on a higher level.

Reply:
I see your point, but that is probably something we should tackle with upcoming PRs since there are multiple issues to discuss (as of now, `filters` are not capable of estimation cardinalities, now they will be with CQFs and histograms; what should be the interface for that, etc.).

Reply:
Mmmh, I don't feel to good about merging PRs with "This needs to be addressed in the future" issues. At least, we should open an issue and make this a priority.

Reply:
Please coordinate with @timzimmermann. If I understand the PRs correctly, their data structures can both answer the question "Is it possible that this element exists". Thus, their interface should be similar.

Reply:
How about calling it `can_contain()`?
`contains()` sounds like the answer is always accurate.

Reply:
I think `does_not_contain` would be even more descriptive and less likely to be misused by somebody that is not particularly interested in pruning.

Reply:
I think `might_contain` is best here. Short, unambigious and accurate.

Reply:
I think out of the proposed, `does_not_contain()` is most clear.
In any case, the docstring will have to describe the exact meaning of true/false anyways, because no short name will be completely descriptive of what the method tells the caller.